{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tools\n",
    "import visualization_2D as vis2D\n",
    "import visualization_3D as vis3D\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.graphics.gofplots import qqplot_2samples\n",
    "from scipy.stats import ks_2samp\n",
    "from scipy.stats import chi2_contingency\n",
    "import matplotlib.pyplot as plt\n",
    "import random as rd\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as px\n",
    "import os\n",
    "import sqlite3\n",
    "from intermine.webservice import Service\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# I- TRN download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "service = Service(\"https://yeastmine.yeastgenome.org/yeastmine/service\")\n",
    "query = service.new_query(\"Gene\")\n",
    "query.add_constraint(\"regulatoryRegions\", \"TFBindingSite\")\n",
    "query.add_view(\n",
    "    \"regulatoryRegions.regulator.symbol\",\n",
    "    \"regulatoryRegions.regulator.secondaryIdentifier\", \"symbol\",\n",
    "    \"secondaryIdentifier\", \"regulatoryRegions.regEvidence.ontologyTerm.name\",\n",
    "    \"regulatoryRegions.regEvidence.ontologyTerm.identifier\",\n",
    "    \"regulatoryRegions.experimentCondition\",\n",
    "    \"regulatoryRegions.strainBackground\",\n",
    "    \"regulatoryRegions.regulationDirection\",\n",
    "    \"regulatoryRegions.publications.pubMedId\", \"regulatoryRegions.datasource\",\n",
    "    \"regulatoryRegions.annotationType\"\n",
    ")\n",
    "query.add_constraint(\"regulatoryRegions.regulator\", \"IN\", \"ALL_Verified_Uncharacterized_Dubious_ORFs\", code=\"A\")\n",
    "query.add_constraint(\"regulatoryRegions.strainBackground\", \"=\", \"S288c\", code=\"B\")\n",
    "query.add_constraint(\"regulatoryRegions.regEvidence.ontologyTerm.name\", \"ONE OF\", \n",
    "                     [\"chromatin immunoprecipitation-chip evidence\", \n",
    "                      \"chromatin immunoprecipitation- exonuclease evidence\", \n",
    "                      \"northern blot evidence\", \n",
    "                      \"western blot evidence\", \n",
    "                      \"chromatin immunoprecipitation-qPCR evidence\", \n",
    "                      \"beta-galactosidase reporter gene assay evidence\", \n",
    "                      \"chromatin immunoprecipitation-PCR evidence\", \n",
    "                      \"co-immunoprecipitation evidence\", \n",
    "                      \"chromatin immunoprecipitation evidence\", \n",
    "                      \"protein expression level evidence based on western blot\"], code=\"C\")\n",
    "\n",
    "# Uncomment and edit the code below to specify your own custom logic:\n",
    "# query.set_logic(\"B and A and C and C\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "TRN = pd.DataFrame(columns=['TF', 'TG'])\n",
    "\n",
    "for row in query.rows():\n",
    "    TRN = TRN.append({'TF': row[\"regulatoryRegions.regulator.secondaryIdentifier\"],\n",
    "                      'TG': row[\"secondaryIdentifier\"]}, ignore_index=True)\n",
    "\n",
    "display(TRN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# II- Targets lists creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "list_TF = TRN[\"TF\"].unique()\n",
    "print(f\"Number of TF: {len(list_TF)}\")\n",
    "\n",
    "for TF in list_TF :\n",
    "    TG = TRN[TRN[\"TF\"] == TF][\"TG\"]\n",
    "    TG.to_csv(f\"../results/V2/TF_target_TRN/{TF}_{len(TG)}_targets.csv\",\n",
    "              index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# III- Cumulative distribution function and distribution histogram"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1) For 3D distances between targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_edges_list(gene_list, edges_list, feature_name):\n",
    "    \n",
    "    # Add SGDID\n",
    "    feature_name = feature_name.merge(genes_list, left_on = \"Feature_name\", right_on = genes_list.columns[0])\n",
    "    \n",
    "    # Extract distances for selected genes list\n",
    "    edges_list_select = edges_list[edges_list[\"Primary_SGDID\"].isin(feature_name[\"Primary_SGDID\"])]\n",
    "    edges_list_select = edges_list_select[edges_list_select[\"Primary_SGDID_bis\"].isin(feature_name[\"Primary_SGDID\"])]\n",
    "    edges_list_select.index = range(1, len(edges_list_select) + 1)\n",
    "    \n",
    "    return edges_list_select\n",
    "\n",
    "\n",
    "def distri(genes_list, edges_list, feature_name):\n",
    "    \n",
    "    sample_number = 20\n",
    "    bin_number = 70\n",
    "    \n",
    "    edges_list_select = get_edges_list(genes_list, edges_list, feature_name)\n",
    "    x = list(edges_list_select[\"3D_distances\"])\n",
    "    H, X1 = np.histogram(x, bins = bin_number)\n",
    "    F1 = np.cumsum(H)/len(x)\n",
    "    \n",
    "    H2 = np.zeros(bin_number)\n",
    "    rd_x = []\n",
    "    for i in range(0, sample_number):\n",
    "        edges_list_random = edges_list.sample(n=len(edges_list_select))\n",
    "        rd_x_tmp = list(edges_list_random[\"3D_distances\"])\n",
    "        H_temp, X_temp = np.histogram(rd_x_tmp, bins = bin_number)\n",
    "        rd_x += rd_x_tmp\n",
    "        H2 += H_temp\n",
    "    H2 = H2/sample_number\n",
    "    F2 = np.cumsum(H2)/sum(H2)\n",
    "    \n",
    "    fig, ax = plt.subplots()\n",
    "    ax.hist(X1[:-1], X1, weights=H2, color=\"#5767FF\", alpha=0.3, label=\"Random targets\")\n",
    "    ax.set_xlim(0, 200)\n",
    "    plt.xlabel(\"3D distances\", size = 16)\n",
    "    ax.hist(x, bins=bin_number, color=\"#FA3824\", alpha=0.3, label=\"TF's targets\")\n",
    "    plt.ylabel(\"Count\", size = 16)\n",
    "    ax2=ax.twinx()\n",
    "    ax2.plot(X1[:-1], F1, label=\"CDF (TF's targets)\", color = \"#FA3824\")\n",
    "    ax2.plot(X1[:-1], F2, label=\"CDF (random)\", color = \"#5767FF\")\n",
    "    ax.legend(bbox_to_anchor = (0.6, 0.9), loc=\"upper left\")\n",
    "    ax2.legend(bbox_to_anchor = (0.6, 0.7), loc=\"upper left\")\n",
    "    \n",
    "    \n",
    "\n",
    "    # Compute Kolmogorov-Smirnov test\n",
    "    # https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ks_2samp.html\n",
    "    if x != []:\n",
    "    #if False:\n",
    "        ks_result = ks_2samp(x, rd_x)\n",
    "    else:\n",
    "        ks_result = None\n",
    "    \n",
    "    return fig, ks_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "adjacency_matrix = pd.read_parquet(\"../dashboard/static/adjacency_matrix_V4.parquet.gzip\", engine='pyarrow')\n",
    "edges_list = adjacency_matrix.stack().dropna().reset_index()\n",
    "edges_list.rename(columns = {0: \"3D_distances\"}, inplace = True)\n",
    "edges_list = edges_list.sort_values(by=\"Primary_SGDID\")\n",
    "edges_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sql_query = \\\n",
    "\"\"\"SELECT Primary_SGDID, Standard_gene_name, Chromosome, Feature_name, Strand, Stop_coordinate, Start_coordinate, Description\n",
    "FROM SGD_features\n",
    "\"\"\"\n",
    "\n",
    "# Get all features for all gene\n",
    "feature_name = tools.get_locus_info(\"../SCERE.db\", sql_query)\n",
    "feature_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#test on one FT\n",
    "genes_list = pd.read_csv('../results/V2/TF_target_TRN/YPR133C_122_targets.csv', sep=',', header = [0])\n",
    "files_names = os.listdir(\"../results/V2/TF_target_TRN\")\n",
    "file_name = 'YPR133C_122_targets.csv'\n",
    "\n",
    "res = distri(genes_list, edges_list, feature_name)\n",
    "\n",
    "res[0].savefig(f\"../results/V2/distri/{file_name}.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "files_names = os.listdir(\"../results/V2/TF_target_TRN\")\n",
    "results = pd.DataFrame(columns = [\"TF\", \"TF_name\", \"description\", \"targets_number\", \"KS_stat\", \"KS_pvalue\"])\n",
    "\n",
    "for file_name in files_names:\n",
    "    print(file_name)\n",
    "    FT_name = file_name.split('_')\n",
    "    FT_name = FT_name[0]\n",
    "    genes_list = pd.read_csv('../results/V2/TF_target_TRN/' + file_name, sep=',', header = [0])\n",
    "    res = distri(genes_list, edges_list, feature_name)\n",
    "    standard_name = feature_name[\"Standard_gene_name\"][feature_name[\"Feature_name\"] == FT_name].values[0]\n",
    "    description = feature_name[\"Description\"][feature_name[\"Feature_name\"] == FT_name].values[0]\n",
    "    \n",
    "    if res[1] != None:\n",
    "        results = results.append({\"TF\": FT_name, \n",
    "                                  \"TF_name\": standard_name, \n",
    "                                  \"description\": description, \n",
    "                                  \"targets_number\": len(genes_list), \n",
    "                                  \"KS_stat\": res[1].statistic, \n",
    "                                  \"KS_pvalue\": res[1].pvalue}, \n",
    "                                 ignore_index=True)\n",
    "    \n",
    "    else:\n",
    "        results = results.append({\"TF\": FT_name, \n",
    "                                  \"TF_name\": standard_name, \n",
    "                                  \"description\": description, \n",
    "                                  \"targets_number\": len(genes_list), \n",
    "                                  \"KS_stat\": None, \n",
    "                                  \"KS_pvalue\": None}, \n",
    "                                 ignore_index=True)\n",
    "    \n",
    "    res[0].savefig(f\"../results/V2/distri/{file_name}.png\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results.sort_values(by =[\"KS_stat\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.histogram(results, x=\"KS_pvalue\", nbins = 10)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2) For 2D chromosomal positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Primary_SGDID</th>\n",
       "      <th>Standard_gene_name</th>\n",
       "      <th>Feature_name</th>\n",
       "      <th>Description</th>\n",
       "      <th>Start_coordinate</th>\n",
       "      <th>Stop_coordinate</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S000028888</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Terminal telomeric repeats on the left arm of ...</td>\n",
       "      <td>34</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S000028892</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Terminal telomeric repeats on the left arm of ...</td>\n",
       "      <td>34</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S000028919</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Terminal telomeric repeats on the left arm of ...</td>\n",
       "      <td>51</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S000001826</td>\n",
       "      <td></td>\n",
       "      <td>YFL068W</td>\n",
       "      <td>Putative protein of unknown function; SWAT-GFP...</td>\n",
       "      <td>53</td>\n",
       "      <td>535</td>\n",
       "      <td>6</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S000030466</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>53</td>\n",
       "      <td>535</td>\n",
       "      <td>6</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16381</th>\n",
       "      <td>S000035945</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1525523</td>\n",
       "      <td>1525095</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16382</th>\n",
       "      <td>S000002953</td>\n",
       "      <td>YRF1-1</td>\n",
       "      <td>YDR545W</td>\n",
       "      <td>Helicase encoded by the Y' element of subtelom...</td>\n",
       "      <td>1526321</td>\n",
       "      <td>1531711</td>\n",
       "      <td>4</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16383</th>\n",
       "      <td>S000036865</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1526321</td>\n",
       "      <td>1531711</td>\n",
       "      <td>4</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16384</th>\n",
       "      <td>S000028617</td>\n",
       "      <td></td>\n",
       "      <td>YDR545C-A</td>\n",
       "      <td>Dubious open reading frame; unlikely to encode...</td>\n",
       "      <td>1531342</td>\n",
       "      <td>1530863</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16385</th>\n",
       "      <td>S000031458</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1531342</td>\n",
       "      <td>1530863</td>\n",
       "      <td>4</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16026 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Primary_SGDID Standard_gene_name Feature_name  \\\n",
       "0        S000028888                                   \n",
       "1        S000028892                                   \n",
       "3        S000028919                                   \n",
       "4        S000001826                         YFL068W   \n",
       "5        S000030466                                   \n",
       "...             ...                ...          ...   \n",
       "16381    S000035945                                   \n",
       "16382    S000002953             YRF1-1      YDR545W   \n",
       "16383    S000036865                                   \n",
       "16384    S000028617                       YDR545C-A   \n",
       "16385    S000031458                                   \n",
       "\n",
       "                                             Description Start_coordinate  \\\n",
       "0      Terminal telomeric repeats on the left arm of ...               34   \n",
       "1      Terminal telomeric repeats on the left arm of ...               34   \n",
       "3      Terminal telomeric repeats on the left arm of ...               51   \n",
       "4      Putative protein of unknown function; SWAT-GFP...               53   \n",
       "5                                                                      53   \n",
       "...                                                  ...              ...   \n",
       "16381                                                             1525523   \n",
       "16382  Helicase encoded by the Y' element of subtelom...          1526321   \n",
       "16383                                                             1526321   \n",
       "16384  Dubious open reading frame; unlikely to encode...          1531342   \n",
       "16385                                                             1531342   \n",
       "\n",
       "      Stop_coordinate  Chromosome Strand  \n",
       "0                   1           7      C  \n",
       "1                   1           8      C  \n",
       "3                   1          13      C  \n",
       "4                 535           6      W  \n",
       "5                 535           6      W  \n",
       "...               ...         ...    ...  \n",
       "16381         1525095           4      C  \n",
       "16382         1531711           4      W  \n",
       "16383         1531711           4      W  \n",
       "16384         1530863           4      C  \n",
       "16385         1530863           4      C  \n",
       "\n",
       "[16026 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql_query = \\\n",
    "\"\"\"SELECT Primary_SGDID, Standard_gene_name, Feature_name, Description, Start_coordinate, Stop_coordinate, Chromosome, Strand\n",
    "FROM SGD_features\n",
    "ORDER BY Start_coordinate\n",
    "\"\"\"\n",
    "\n",
    "# Get all features for all gene\n",
    "loci = tools.get_locus_info(\"../SCERE.db\", sql_query)\n",
    "loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_chrom_list(gene_list, loci):\n",
    "    \n",
    "    # Add SGDID\n",
    "    chrom_list = loci[loci[\"Feature_name\"].isin(genes_list[\"TG\"])]\n",
    "    \n",
    "    chrom_list.index = range(1, len(chrom_list) + 1)\n",
    "    \n",
    "    return chrom_list\n",
    "\n",
    "\n",
    "def chrom_distri(genes_list, loci):\n",
    "    \n",
    "    sample_number = 20\n",
    "    \n",
    "    chrom_list = get_chrom_list(genes_list, loci)\n",
    "    x = list(chrom_list[\"Chromosome\"])\n",
    "    H, X1 = np.histogram(x, bins = list(range(1, 18)))\n",
    "    \n",
    "    H2 = np.zeros(16)\n",
    "    for i in range(0, sample_number):\n",
    "        chrom_list_random = loci.sample(n=len(chrom_list))\n",
    "        rd_x_tmp = list(chrom_list_random[\"Chromosome\"])\n",
    "        H_temp, X_temp = np.histogram(rd_x_tmp, bins = list(range(1, 18)))\n",
    "        H2 += H_temp\n",
    "    H2 = np.round_(H2/sample_number)\n",
    "    \n",
    "    fig, ax = plt.subplots()\n",
    "    ax.hist(X1[:-1], X1, weights=H2, color=\"#5767FF\", alpha=0.3, label=\"Random targets\")\n",
    "    ax.set_xlim(1, 16)\n",
    "    plt.xlabel(\"Chromosomes\", size = 16)\n",
    "    ax.hist(x, bins=list(range(1, 18)), color=\"#FA3824\", alpha=0.3, label=\"TF's targets\")\n",
    "    plt.ylabel(\"Count\", size = 16)\n",
    "    ax.legend(bbox_to_anchor = (0.6, 0.9), loc=\"upper left\")\n",
    "\n",
    "    # Compute chi square test\n",
    "    #https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html    if x != []:\n",
    "    if ((0 in H) or (0 in H2)) == False:\n",
    "        chi_test = np.array([H, H2])\n",
    "        chi_result = chi2_contingency(chi_test)[:2]\n",
    "    else:\n",
    "        chi_result = (0, 0)\n",
    "        \n",
    "    return fig, chi_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_list = pd.read_csv('../results/TF_target_TRN/YKR034W_2_targets.csv', sep=',', header = [0])\n",
    "files_names = os.listdir(\"../results/TF_target_TRN\")\n",
    "file_name = 'YKR034W_2_targets.csv'\n",
    "\n",
    "res = chrom_distri(genes_list, loci)\n",
    "\n",
    "res[0].show()\n",
    "print(res[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "files_names = os.listdir(\"../results/TF_target_TRN\")\n",
    "results = pd.DataFrame(columns = [\"TF\", \"TF_name\", \"description\", \"targets_number\", \"Chi_stat\", \"Chi_pvalue\"])\n",
    "\n",
    "for file_name in files_names:\n",
    "    print(file_name)\n",
    "    FT_name = file_name.split('_')\n",
    "    FT_name = FT_name[0]\n",
    "    genes_list = pd.read_csv('../results/TF_target_TRN/' + file_name, sep=',', header = [0])\n",
    "    res = chrom_distri(genes_list, loci)\n",
    "    standard_name = loci[\"Standard_gene_name\"][loci[\"Feature_name\"] == FT_name].values[0]\n",
    "    description = loci[\"Description\"][loci[\"Feature_name\"] == FT_name].values[0]\n",
    "    \n",
    "    if res[1] != (0, 0):\n",
    "        results = results.append({\"TF\": FT_name, \n",
    "                                  \"TF_name\": standard_name, \n",
    "                                  \"description\": description, \n",
    "                                  \"targets_number\": len(genes_list), \n",
    "                                  \"Chi_stat\": res[1][0], \n",
    "                                  \"Chi_pvalue\": res[1][1]}, \n",
    "                                 ignore_index=True)\n",
    "    \n",
    "    else:\n",
    "        results = results.append({\"TF\": FT_name, \n",
    "                                  \"TF_name\": standard_name, \n",
    "                                  \"description\": description, \n",
    "                                  \"targets_number\": len(genes_list), \n",
    "                                  \"Chi_stat\": None, \n",
    "                                  \"Chi_pvalue\": None}, \n",
    "                                 ignore_index=True)\n",
    "    \n",
    "    res[0].savefig(f\"../results/chrom_distri/{file_name}.png\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results.sort_values(by =[\"Chi_stat\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\"\"\n",
    "SELECT Primary_SGDID, Feature_name, Feature_type, Start_coordinate, Strand, Chromosome\n",
    "FROM SGD_features\n",
    "\"\"\"\n",
    "\n",
    "cursor = db_connexion.cursor()\n",
    "    \n",
    "# Query database.\n",
    "chrom_info = cursor.execute(query)\n",
    "    \n",
    "# Convert to Pandas dataframe\n",
    "column_names = [column[0] for column in chrom_info.description]\n",
    "chrom_info_df = pd.DataFrame(chrom_info.fetchall(), columns=column_names)\n",
    "    \n",
    "# Select only strands + and -\n",
    "chrom_info_df = chrom_info_df[ (chrom_info_df[\"Strand\"] == \"C\") | (chrom_info_df[\"Strand\"] == \"W\") ]\n",
    "# Remove \"2-micron\" plasmid\n",
    "chrom_info_df = chrom_info_df[ chrom_info_df[\"Chromosome\"] != \"2-micron\" ]\n",
    "# Convert chromosome id to int\n",
    "chrom_info_df[\"Chromosome\"] = chrom_info_df[\"Chromosome\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom_info_df[ chrom_info_df[\"Feature_type\"] != \"CDS\" ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>TG</th>\n",
       "      <th>type</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABF1</td>\n",
       "      <td>COX6</td>\n",
       "      <td>Direct Positive</td>\n",
       "      <td>N/A;N/A;N/A;Stress;Unstressed log-phase growth...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABF1</td>\n",
       "      <td>RPS28B</td>\n",
       "      <td>Direct</td>\n",
       "      <td>N/A;N/A;Stress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABF1</td>\n",
       "      <td>RPO26</td>\n",
       "      <td>Direct</td>\n",
       "      <td>N/A;N/A;N/A;Unstressed log-phase growth (control)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABF1</td>\n",
       "      <td>HIS7</td>\n",
       "      <td>Direct Dual</td>\n",
       "      <td>N/A;In vitro;Unstressed log-phase growth (cont...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABF1</td>\n",
       "      <td>ADE5,7</td>\n",
       "      <td>Direct</td>\n",
       "      <td>N/A;Unstressed log-phase growth (control)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195493</th>\n",
       "      <td>PEX2</td>\n",
       "      <td>PEX2</td>\n",
       "      <td>Indirect Negative</td>\n",
       "      <td>Unstressed log-phase growth (control);Stress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195494</th>\n",
       "      <td>FYV5</td>\n",
       "      <td>ATG14</td>\n",
       "      <td>Indirect Negative</td>\n",
       "      <td>Unstressed log-phase growth (control)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195495</th>\n",
       "      <td>FYV5</td>\n",
       "      <td>ATG9</td>\n",
       "      <td>Indirect Negative</td>\n",
       "      <td>Unstressed log-phase growth (control)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195496</th>\n",
       "      <td>FYV5</td>\n",
       "      <td>ATG32</td>\n",
       "      <td>Indirect Negative</td>\n",
       "      <td>Unstressed log-phase growth (control)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195497</th>\n",
       "      <td>FYV5</td>\n",
       "      <td>ATG8</td>\n",
       "      <td>Indirect Negative</td>\n",
       "      <td>Unstressed log-phase growth (control)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>195498 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TF      TG               type  \\\n",
       "0       ABF1    COX6    Direct Positive   \n",
       "1       ABF1  RPS28B             Direct   \n",
       "2       ABF1   RPO26             Direct   \n",
       "3       ABF1    HIS7        Direct Dual   \n",
       "4       ABF1  ADE5,7             Direct   \n",
       "...      ...     ...                ...   \n",
       "195493  PEX2    PEX2  Indirect Negative   \n",
       "195494  FYV5   ATG14  Indirect Negative   \n",
       "195495  FYV5    ATG9  Indirect Negative   \n",
       "195496  FYV5   ATG32  Indirect Negative   \n",
       "195497  FYV5    ATG8  Indirect Negative   \n",
       "\n",
       "                                              description  \n",
       "0       N/A;N/A;N/A;Stress;Unstressed log-phase growth...  \n",
       "1                                          N/A;N/A;Stress  \n",
       "2       N/A;N/A;N/A;Unstressed log-phase growth (control)  \n",
       "3       N/A;In vitro;Unstressed log-phase growth (cont...  \n",
       "4               N/A;Unstressed log-phase growth (control)  \n",
       "...                                                   ...  \n",
       "195493       Unstressed log-phase growth (control);Stress  \n",
       "195494              Unstressed log-phase growth (control)  \n",
       "195495              Unstressed log-phase growth (control)  \n",
       "195496              Unstressed log-phase growth (control)  \n",
       "195497              Unstressed log-phase growth (control)  \n",
       "\n",
       "[195498 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Yeast_TRN = pd.read_csv('../results/yeast2019-full-conds-net.csv', sep='\\t', header = None)\n",
    "Yeast_TRN.columns = [\"TF\", \"TG\", \"type\", \"description\"]\n",
    "Yeast_TRN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6886"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Yeast_TRN[1].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_Yeast_TRN = Yeast_TRN[Yeast_TRN.type.isin(['Direct Positive', 'Direct', 'Direct Dual', 'Direct Negative'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6475"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(selected_Yeast_TRN.TG.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
